juju/juju
# [JUJU-121] Adds support for auto create of instance profiles.

This commit adds support instance profiles auto creation. Including the associated role and policy needed to scope permissions correctly.

Accompanying this PR is also a discourse doc on the permissions Juju requires when talking with AWS.

## QA steps

```sh
juju bootstrap --bootstrap-constraints="instance-role=auto" aws/ap-southeast-2 test-tlm-controller
```

You then need to confirm with the aws cli that the associated instance profile elements were created. Make sure that you see a role with the same here attached to the instance profile.

```sh
aws iam get-instance-profile --instance-profile-name juju-controller-test-tlm-controller
```

Add some machines to Juju and deploy a charm with storage to confirm permissions are operating correctly.

Check HA works

```
$ juju enable-ha
juju enable-ha
maintaining machines: 0
adding machines: 1, 2
$ juju status
Model       Controller     Cloud/Region        Version   SLA          Timestamp
controller  test-tlm-controller aws/ap-southeast-2  2.9.20.1  unsupported  09:34:28+10:00

Machine  State    DNS           Inst id              Series  AZ               Message
0        started  13.211.59.65  i-f00f00f00f00f00f00  focal   ap-southeast-2b  running
1        pending                pending              focal                    attaching aws instance profile arn:aws:iam::123456789:instance-profile/juju-controller-test-tlm-controller
2        pending                pending              focal                    attaching aws instance profile arn:aws:iam::123456789:instance-profile/juju-controller-test-tlm-controller
```

Check destroy-controller works
```
$ juju destroy-controller test-tlm-controller --destroy-all-models
```

## Documentation changes

Permission list: https://discourse.charmhub.io/t/juju-aws-permissions/5307
Doc Link: https://discourse.charmhub.io/t/using-aws-instance-profiles-with-juju-2-9/5185/3

## Bug reference

N/A
